hls4ml/templates/vivado/nnet_utils/nnet_array.h

```diff
@@ -12,24 +12,24 @@ struct transpose_config {
     static constexpr unsigned perm[3] = {2, 0, 1};
 };
 
-template<class data_T, typename CONFIG_T>
+template<class data_T, class res_T, typename CONFIG_T>
 void transpose_2d(
     data_T data[CONFIG_T::height * CONFIG_T::width],
-    data_T data_t[CONFIG_T::height * CONFIG_T::width]
+    res_T data_t[CONFIG_T::height * CONFIG_T::width]
```

Comment:
What's the reason for introducing a data type for the output? `Transpose` is one of those layers that shouldn't be able to change the output precision. I think [we're still safe](https://github.com/fastmachinelearning/hls4ml/blob/master/hls4ml/model/hls_layers.py#L1569), but forcing it at the HLS level by using the same type doesn't seem like a bad idea.

Reply:
The only reason I wanted to have this here is to have a consistent interface with the streaming version.

For the streaming version, we use `data_T::size` and `res_T::size` because the incoming streams and the outgoing streams should have different depths.

So it makes the python frontend code cleaner because this line works for both:
https://github.com/fastmachinelearning/hls4ml/blob/1da1c719900793eccdf04bf4a263ab5a3de47d8d/hls4ml/templates/vivado_template.py#L367

Reply:
Yeah that makes sense, and there is protection against them being different in the Python.